### ISSUE TITLE
`@JsonProperty` not serializing field names properly on `@JsonCreator` in Record

### ISSUE THREAD

**Issue by Incara** [FasterXML/jackson-databind#4452]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Serialization of java records do not appear to be honoring `@JsonProperty` names in `@JsonCreator` constructors. 

### Version Information

2.15.x+

### Reproduction

A test below fails for me and I expect it to pass.

```java
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

public class RecordTest {

    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    public record TestObject(String testFieldName, Integer testOtherField) {

        @JsonCreator
        public TestObject(@JsonProperty ("strField") String testFieldName,
                          @JsonProperty ("intField") Integer testOtherField) {
            this.testFieldName = testFieldName;
            this.testOtherField = testOtherField;
        }
    }

    @Test
    public void test() throws JsonProcessingException {
        final String result = OBJECT_MAPPER.writeValueAsString(new TestObject("test", 1));
        assertTrue(result.contains("strField"));
    }
}
```

### Expected behavior

The field is renamed to what is specified in the `@JsonProperty` annotation in the record's `@JsonCreator`

### Additional context

We encountered this upgrading from 2.14.x up to 2.16.x, so this functionality was working at that earlier version (perhaps before records were really addressed). 

Perhaps we have versions incorrect between jackson modules and libraries, but we haven't found the exact mismatch yet.

**Comment by Incara** [comment]
Yeah, I think we've figured out this test specifically doesn't work on any version. We're not sure why it works for us with our application's dependency combination. I don't know if it's something about the parameter names module or properties on the spring `ObjectMapper` injected at runtime that allows it to be possible. 

Is this something that is intended to be supported or is it a requirement that `@JsonProperty` annotations must appear in the record declaration?

**Comment by Incara** [comment]
Much appreciated, this helps with analysis for sure. I think that narrows this down to something changed between `2.15.4` and `2.16.0` and now it doesn't seem like `@JsonProperty` renames are honored on `@JsonCreator` in `record`s. The below works in `2.15.4` but fails in `2.16.0`:

```java
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;

public class RecordTest {

    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
            .registerModule(new ParameterNamesModule()).findAndRegisterModules();

    public record TestObject(String testFieldName, Integer testOtherField) {

        @JsonCreator
        public TestObject(@JsonProperty ("strField") String testFieldName,
                          @JsonProperty ("intField") Integer testOtherField) {
            this.testFieldName = testFieldName;
            this.testOtherField = testOtherField;
        }
    }

    @Test
    public void test() throws JsonProcessingException {
        final String result = OBJECT_MAPPER.writeValueAsString(new TestObject("test", 1));
        assertTrue(result.contains("strField"));
    }
}
```

For now, our solution is to eliminate our `@JsonCreator`s on `record`s in favor using `@JsonProperty` since there is larger support in jackson for `record`s since we last updated.

I don't know if there is supposed to be support for the above test case from the repo maintainers, but I will also stop and let them comment.

**Comment by JooHyukKim** [comment]
So as per [release note](https://github.com/FasterXML/jackson-databind/blob/2.18/release-notes/CREDITS-2.x) there seem to be these three PRs --#3906, #3992, #4175 -- that are titled with the word `record` in it.

**Comment by yihtserns** [comment]
> ...titled with the word `record` in it.

I don't think this thing is Record-specific.

**Comment by JooHyukKim** [comment]
> The below works in `2.15.4` but fails in `2.16.0`:

I guess `ParameterNamesModule` version is same with `jackson-databind`?

**Comment by JooHyukKim** [comment]
Because of project's current on-going milestone aka `Property Introspection Rewrite` (also mentioned in [2.17](https://groups.google.com/d/msgid/jackson-dev/CAL4a10gtvE9_X78DtEBpyA%3DV_0WjrxgpzJ9kUif_aBLKRu-noQ%40mail.gmail.com)), we neither can have have fix any time soon nor confidently specify how record should behave.

**Comment by cowtowncoder** [comment]
FWTW, if not Record-specific, reproduction without Record would be great (since it's likely in shared code b/w POJOs, Records).

**Comment by cowtowncoder** [comment]
Looking at the original case, it does seem like a flaw: `@JsonCreator`-annotated constructor should be used as the primary one, and any property name changes should be used from that.

And while fix may or may not be easy, a PR for reproduction, would be great.
It seems to be this would be Record-specific thing, fwtw, as fully annotated case


> > ...titled with the word `record` in it.
> 
> I don't think this thing is Record-specific.

Looking at the original reproduction, it would seem it is?
Or why do you think it isn't?

**Comment by cowtowncoder** [comment]
Ok, at any rate: a PR for failing test case (under `src/test-jdk17/java/..../failing`) would be great, either against 2.17 or 2.18 branch.

**Comment by JooHyukKim** [comment]
> Ok, at any rate: a PR for failing test case (under `src/test-jdk17/java/..../failing`) would be great, either against 2.17 or 2.18 branch.

Most test cases here make use of `ParameterNamesModule`, so it seems not possible?

**Comment by yihtserns** [comment]
@JooHyukKim I think this can help: https://github.com/FasterXML/jackson-databind/blob/2.18/src/test-jdk17/java/com/fasterxml/jackson/databind/records/Jdk8ConstructorParameterNameAnnotationIntrospector.java

**Comment by cowtowncoder** [comment]
Ugh. I missed `ParameterNamesModule` reference since that wasn't in the original description.

Correct, tests here cannot use it.

There are a few tests tho that implement replica, `AnnotationIntrospector` that provides bogus names for testing. So it is possible, if tedious, to do that (beside one that @yihtserns referenced).

**Comment by JooHyukKim** [comment]
Filed PR #4477. It does fail from 2.16 version and on.

**Comment by cowtowncoder** [comment]
Ok, I know what is causing this: before #4515, Constructor properties are not resolved for Record serialization (for some reason). It is likely this can be fixed ones #4515 fix is complete enough.
